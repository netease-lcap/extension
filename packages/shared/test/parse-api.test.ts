import { expect, test } from 'vitest';
import parseComponentApi from '../src/transforms/parse-component-api';

const apiCode = `
/// <reference types="@nasl/types" />
namespace extensions.vue_library_example.viewComponents {
  const { Component, Prop, ViewComponent, Slot, Method, Param, Event, ViewComponentOptions } = nasl.ui;

  @ExtensionComponent({
    type: 'both',
    ideusage: {
      idetype: 'element',
    }
  })
  @Component({
    title: '二维码查看',
    description: '二维码查看组件，可生成和展示二维码',
  })
  export class QrCode extends ViewComponent {
    constructor(options?: Partial<QrCodeOptions>) {
      super();
    }
  }

  export class QrCodeOptions extends ViewComponentOptions {
    @Prop({
      title: '扫描内容',
      description: '二维码扫描后的文本或链接',
      group: '主要属性',
      setter: {
        concept: 'InputSetter'
      }
    })
    value: nasl.core.String = 'https://example.com';

    @Prop({
      title: '中心图标',
      description: '二维码中心的图片地址',
      group: '主要属性',
      setter: {
        concept: 'ImageSetter'
      }
    })
    icon: nasl.core.String = '';

    @Prop({
      title: '图标大小',
      description: '二维码中心的图片大小',
      group: '主要属性',
      setter: {
        concept: 'EnumSelectSetter',
        options: [{ title: '小' }, { title: '中' }, { title: '大' }]
      }
    })
    iconSize: 'small' | 'medium' | 'large' = 'medium';

    @Prop({
      title: '二维码颜色',
      description: '二维码的颜色，支持十六进制颜色值',
      group: '样式属性',
      setter: {
        concept: 'InputSetter'
      }
    })
    color: nasl.core.String = '#000000';

    @Prop({
      title: '背景颜色',
      description: '二维码的背景颜色，支持十六进制颜色值',
      group: '样式属性',
      setter: {
        concept: 'InputSetter'
      }
    })
    backgroundColor: nasl.core.String = '#FFFFFF';

    @Prop({
      title: '支持放大',
      description: '是否支持点击放大查看二维码',
      group: '交互属性',
      setter: {
        concept: 'SwitchSetter'
      }
    })
    zoomable: nasl.core.Boolean = false;
  }
}
`;

test('parse component api', () => {
  const result = parseComponentApi(apiCode, 'vue2');
  expect(result).toMatchSnapshot();
});

