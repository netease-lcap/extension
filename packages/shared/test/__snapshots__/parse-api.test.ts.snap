// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`parse component api 1`] = `
[
  {
    "blocks": [],
    "children": [],
    "concept": "ViewComponentDeclaration",
    "description": "二维码查看组件，可生成和展示二维码",
    "events": [],
    "group": "",
    "icon": "",
    "ideusage": {
      "idetype": "element",
    },
    "kebabName": "qr-code",
    "methods": [],
    "name": "QrCode",
    "props": [
      {
        "bindHide": undefined,
        "bindOpen": undefined,
        "concept": "PropDeclaration",
        "defaultValue": {
          "concept": "DefaultValue",
          "expression": {
            "concept": "StringLiteral",
            "value": "https://example.com",
          },
          "playground": [],
        },
        "description": "二维码扫描后的文本或链接",
        "group": "主要属性",
        "name": "value",
        "setter": {
          "concept": "InputSetter",
        },
        "sync": undefined,
        "tabKind": "property",
        "title": "扫描内容",
        "tsType": "nasl.core.String",
      },
      {
        "bindHide": undefined,
        "bindOpen": undefined,
        "concept": "PropDeclaration",
        "defaultValue": {
          "concept": "DefaultValue",
          "expression": {
            "concept": "StringLiteral",
            "value": "",
          },
          "playground": [],
        },
        "description": "二维码中心的图片地址",
        "group": "主要属性",
        "name": "icon",
        "setter": {
          "concept": "ImageSetter",
        },
        "sync": undefined,
        "tabKind": "property",
        "title": "中心图标",
        "tsType": "nasl.core.String",
      },
      {
        "bindHide": undefined,
        "bindOpen": undefined,
        "concept": "PropDeclaration",
        "defaultValue": {
          "concept": "DefaultValue",
          "expression": {
            "concept": "StringLiteral",
            "value": "medium",
          },
          "playground": [],
        },
        "description": "二维码中心的图片大小",
        "group": "主要属性",
        "name": "iconSize",
        "setter": {
          "concept": "EnumSelectSetter",
          "options": [
            {
              "title": "小",
              "value": "small",
            },
            {
              "title": "中",
              "value": "medium",
            },
            {
              "title": "大",
              "value": "large",
            },
          ],
        },
        "sync": undefined,
        "tabKind": "property",
        "title": "图标大小",
        "tsType": "'small' | 'medium' | 'large'",
      },
      {
        "bindHide": undefined,
        "bindOpen": undefined,
        "concept": "PropDeclaration",
        "defaultValue": {
          "concept": "DefaultValue",
          "expression": {
            "concept": "StringLiteral",
            "value": "#000000",
          },
          "playground": [],
        },
        "description": "二维码的颜色，支持十六进制颜色值",
        "group": "样式属性",
        "name": "color",
        "setter": {
          "concept": "InputSetter",
        },
        "sync": undefined,
        "tabKind": "property",
        "title": "二维码颜色",
        "tsType": "nasl.core.String",
      },
      {
        "bindHide": undefined,
        "bindOpen": undefined,
        "concept": "PropDeclaration",
        "defaultValue": {
          "concept": "DefaultValue",
          "expression": {
            "concept": "StringLiteral",
            "value": "#FFFFFF",
          },
          "playground": [],
        },
        "description": "二维码的背景颜色，支持十六进制颜色值",
        "group": "样式属性",
        "name": "backgroundColor",
        "setter": {
          "concept": "InputSetter",
        },
        "sync": undefined,
        "tabKind": "property",
        "title": "背景颜色",
        "tsType": "nasl.core.String",
      },
      {
        "bindHide": undefined,
        "bindOpen": undefined,
        "concept": "PropDeclaration",
        "defaultValue": {
          "concept": "DefaultValue",
          "expression": {
            "concept": "BooleanLiteral",
            "value": "false",
          },
          "playground": [],
        },
        "description": "是否支持点击放大查看二维码",
        "group": "交互属性",
        "name": "zoomable",
        "setter": {
          "concept": "SwitchSetter",
        },
        "sync": undefined,
        "tabKind": "property",
        "title": "支持放大",
        "tsType": "nasl.core.Boolean",
      },
    ],
    "readableProps": [],
    "slots": [],
    "themeVariables": [],
    "title": "二维码查看",
    "tsTypeParams": "",
    "type": "both",
  },
]
`;
